lib/compilation_engine.rb

```diff
@@ -57,52 +68,192 @@ def compile_subroutine
   def compile_parameter_list
     @output.puts("<parameterList>")
 
-    advance
-    until @tokenizer.token_type == :SYMBOL && @tokenizer.symbol == ")"
-      output_token
-      advance
+    unless is_symbol_token_and_equal?(")")
+      output_token # type
+      output_token # varName
+
+      while is_symbol_token_and_equal?(",")
+        output_token # ,
+        output_token # type
+        output_token # varName
+      end
     end
 
     @output.puts("</parameterList>")
   end
 
   def compile_var_dec
     @output.puts("<varDec>")
-    output_token
-    advance_and_output_token
-    advance_and_output_token
+    output_token # var
+    output_token # type
+    output_token # varName
 
-    advance
-    until @tokenizer.token_type == :SYMBOL && @tokenizer.symbol == ";"
-      output_token
-      advance_and_output_token
-      advance
+    while is_symbol_token_and_equal?(",")
+      output_token # ,
+      output_token # varName
     end
 
-    output_token
+    output_token # ;
+
     @output.puts("</varDec>")
   end
 
+  def compile_statements
+    @output.puts("<statements>")
+
+    until !@tokenizer.has_more_tokens? || is_symbol_token_and_equal?("}")
```

Comment:
This condition sticks out as being clumsier and harder to understand than the earlier ones. I remember how we got here, but I think we could now get away with just…

```suggestion
    while @tokenizer.token_type == :KEYWORD
```

…which we could write even more nicely as `while keyword?` (or whatever) if we had a helper for it. (Bonus: checking that `@tokenizer.token_type == :KEYWORD` would also make it more technically correct for us to call `@tokenizer.key_word` on the next line.)